lib/Dialect/Ibis/Transforms/IbisCallPrep.cpp

```diff
@@ -0,0 +1,270 @@
+                              {FlatSymbolRefAttr::get(method)});
+  }
+};
+
+CallPrepPrecomputed::CallPrepPrecomputed(ModuleOp mod) {
+  auto *ctxt = mod.getContext();
+
+  // Populate the class-symbol lookup table.
+  for (auto cls : mod.getOps<ClassOp>())
+    classSymbols[cls.getSymNameAttr()] = cls;
+
+  for (auto cls : mod.getOps<ClassOp>()) {
+    // Compute new argument types for each method.
+    for (auto method : cls.getOps<MethodOp>()) {
+
+      // Create the struct type.
+      SmallVector<hw::StructType::FieldInfo> argFields;
+      for (auto [argName, argType] :
+           llvm::zip(method.getArgNamesAttr().getAsRange<StringAttr>(),
+                     method.getArgumentTypes()))
+        argFields.push_back({argName, argType});
+      auto argStruct = hw::StructType::get(ctxt, argFields);
+
+      // Later we're gonna want the block locations, so compute a fused location
+      // and store it.
+      Location argLoc = UnknownLoc::get(ctxt);
+      if (method->getNumRegions() > 0) {
+        SmallVector<Location> argLocs;
+        Block *body = &method.getBody().front();
+        for (auto arg : body->getArguments())
+          argLocs.push_back(arg.getLoc());
+        argLoc = FusedLoc::get(ctxt, argLocs);
+      }
+
+      // Add both to the lookup table.
+      argTypes.insert(
+          std::make_pair(getSymbol(method), std::make_pair(argStruct, argLoc)));
+    }
+
+    // Populate the instances table.
+    for (auto inst : cls.getOps<InstanceOp>()) {
+      auto clsEntry = classSymbols.find(inst.getClassNameAttr().getAttr());
+      assert(clsEntry != classSymbols.end() &&
+             "class being instantiated doesn't exist");
+      instanceMap[std::make_pair(cls, inst.getSymNameAttr())] =
+          clsEntry->second;
+    }
+  }
+}
+
+SymbolRefAttr
+CallPrepPrecomputed::resolveInstancePath(Operation *scope,
+                                         SymbolRefAttr path) const {
+  auto cls = scope->getParentOfType<ClassOp>();
+  assert(cls && "scope outside of ibis class");
+
+  // SymbolRefAttr is rather silly. The start of the path is root reference...
+  cls = lookupNext(cls, path.getRootReference());
+  if (!cls)
+    return {};
+
+  // ... then the rest are the nested references. The last one is the function
+  // name rather than an instance.
+  for (auto instSym : path.getNestedReferences().drop_back()) {
+    cls = lookupNext(cls, instSym.getAttr());
+    if (!cls)
+      return {};
+  }
+
+  // The last one is the function symbol.
+  return SymbolRefAttr::get(cls.getSymNameAttr(),
+                            {FlatSymbolRefAttr::get(path.getLeafReference())});
+}
+
+namespace {
+/// For each CallOp, the corresponding method signature will have changed. Pack
+/// all the operands into a struct.
+struct MergeCallArgs : public OpConversionPattern<CallOp> {
+  MergeCallArgs(MLIRContext *ctxt, const CallPrepPrecomputed &info)
+      : OpConversionPattern(ctxt, 1), info(info) {}
```

Comment:
Why give this (and the pattern below) an explicit benefit value and not just go with whatever's default? (`OpConversionPattern(ctxt)`).

Reply:
VSCode was insisting I needed it. But it was wrong.